### PR TITLE
IB/MD: added config for custom PCI BW

### PR DESCRIPTION
**Pull request by hoopoepg**
- added config for custom PCI bandwidth. this config should be used
  to set actual limit for bandwidth in case when PCI can't provide
  enough bandwidth

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8790/ for details (Mellanox internal link).

File: src/uct/ib/base/ib_md.c [hunk 1]
```diff
@@ -1402,6 +1411,20 @@ uct_ib_md_parse_subnet_prefix(const char *subnet_prefix_str,
...
 }
 
+static double uct_ib_md_pci_bw(const uct_ib_md_config_t *md_config,
+                               struct ibv_device *ib_device)
+{
+    unsigned i;
+
+    for (i = 0; i < md_config->pci_bw.count; i++) {
+        if (!strcmp(ib_device->name, md_config->pci_bw.device[i].name)) {
+            return md_config->pci_bw.device[i].bw * 0.9846; /* ~1.54% lost for PCI Gen3 */
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: pls add formula which was used to calculate 0.9846 according to gen3 spec (e.g 64/66*128/148... etc)

Comment (hoopoepg) [reply src/uct/ib/base/ib_md.c]: Info from this article: https://www.tested.com/tech/457440-theoretical-vs-actual-bandwidth-pci-express-and-thunderbolt/

Comment (hoopoepg) [reply src/uct/ib/base/ib_md.c]: added explanation

File: src/uct/ib/base/ib_md.c [hunk 2]
```diff
@@ -1418,7 +1418,10 @@ static double uct_ib_md_pci_bw(const uct_ib_md_config_t *md_config,
 
     for (i = 0; i < md_config->pci_bw.count; i++) {
         if (!strcmp(ib_device->name, md_config->pci_bw.device[i].name)) {
-            return md_config->pci_bw.device[i].bw * 0.9846; /* ~1.54% lost for PCI Gen3 */
+            return md_config->pci_bw.device[i].bw * 0.9846; /* ~1.54% lost for PCI Gen3. See below */
+           /* PCI Express 3.0 uses 128b/130b encoding scheme: 128 bits of data + 2 bits for *
+            * correctness. It means that ~1.54% (1 - 128/130) of data is used for PCI       *
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: this calculation is incomplete, for 8x Gen3 should be ~108 Gb/s
also - what about Gen1/Gen2?

Comment (yosefe) [reply src/uct/ib/base/ib_md.c]: pls do it like this: https://github.com/openucx/ucx/blob/95031af7e4943411b7d23327db0b24bc9c11908c/src/uct/ib/base/ib_iface.c#L1163

Comment (hoopoepg) [reply src/uct/ib/base/ib_md.c]: Gen1/2 uses 8/10 bits encoding, about 20% overhead, and there is no way to know which gen is used (assuming the most actual for now - Gen3)

are you sure that 8x Gen3 (8Gbs) provides 108 Gbs? not 16x?
note that real speed will depends from memory/PCI/HCA protocol which is magic number...

ok, updated

Comment (yosefe) [reply src/uct/ib/base/ib_md.c]: we can detect gen according to speed of single lane: Gen1: 2.5Gbs, Gen2: 5Gbs, Gen3: 8Gbs
in PCIe there is encoding and above that TLP header + flow control overhead
see https://www.xilinx.com/support/documentation/white_papers/wp350.pdf

Comment (hoopoepg) [reply src/uct/ib/base/ib_md.c]: according offline discussion - multiplexer is removed

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8793/ for details (Mellanox internal link).

File: src/uct/ib/base/ib_md.c [hunk 3]
```diff
@@ -134,7 +134,7 @@ static ucs_config_field_t uct_ib_md_config_table[] = {
 #endif
 
     {"PCI_BW", "",
-     "Maximum bandwidth of PCI bus connected to HCA\n",
+     "Maximum fine-grained bandwidth of PCI bus connected to HCA\n",
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: don't know what is fine-grained.. pls write "effective throughput" or "effective data transfer rate" as in XIlinx doc

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8794/ for details (Mellanox internal link).

File: src/uct/ib/base/ib_md.c (hunk 3 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_md.c]: wording updated

I will squash changes after approve

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8797/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8798/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8802/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8804/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
@yosefe ok to merge?

**Pull request merged**